### DISCUSSION TITLE
autoUpdate() runs the reposition callback twice on init

### DISCUSSION THREAD

**Discussion by claviska** [floating-ui/floating-ui#1837] [Q&A] [answered]
I have some code that controls the positioner and keeps everything anchored. It looks something like this.

```ts
let cleanup;

function start() {
  // Set the initial position
  reposition();

  // Keep it anchored
  cleanup = autoUpdate(anchor, popup, () => reposition());
}

function stop() {
  // Clean everything up
  if (cleanup) {
    cleanup();
    cleanup = undefined;
  }
}

function reposition() {
  // Update the popup's position
  computePosition(...).then(...);
}
```

When I call `start()`, everything works fine except with `autoUpdate()`, the `reposition()` method gets called three times (once by `start()` and twice by the `autoUpdate()` callback) instead of just once.

It's not clear from the [`autoUpdate()` docs](https://floating-ui.com/docs/autoUpdate) whether or not the callback is supposed to be executed right away, and I'm not sure why it's being called a second time. (It happens on a simple page reload with no resizing or any interaction from the user.)

My question is, what's the correct pattern for creating a start/stop controller like this? Ideally, I'd like to see `reposition()` called just once, but I've only been able to achieve that by removing `autoUpdate()`. Is it safe to skip the initial call to `reposition()` since `autoUpdate()` calls it? Is that by design or is it a side effect of the internal resize observer perhaps?

**Comment by atomiks** [comment, marked as answer]
In v0.5.x it was changed to immediately call it, so you don't need to call your update function before. That's why the docs now have an inlined update function, because you don't need to call it yourself.

**Reply by claviska** [reply]
Thanks for clarifying. Do you have any insight as to why it's being called a second time? Is the first time on init and then second a side effect of the resize observer, perhaps?

**Reply by atomiks** [reply]
The ResizeObserver's update is ignored initially. If you console.trace() after the data is resolved in computePosition, it might reveal something
